src/persist/src/file.rs

```diff
@@ -0,0 +1,204 @@
+
+            // Decode val
+            let (len_bytes, remainder) = bytes_remaining.split_at(std::mem::size_of::<u64>());
+            let val_len = u64::from_le_bytes(len_bytes.try_into().unwrap());
+
+            let (val_bytes, remainder) = remainder.split_at(val_len as usize);
+            let val = String::from_utf8(val_bytes.to_vec()).expect("failed to parse utf8");
+            bytes_remaining = remainder;
+
+            // Decode ts and diff
+            let (ts_bytes, remainder) = bytes_remaining.split_at(std::mem::size_of::<u64>());
+            let (diff_bytes, remainder) = remainder.split_at(std::mem::size_of::<i64>());
+            let ts = u64::from_le_bytes(ts_bytes.try_into().unwrap());
+            let diff = i64::from_le_bytes(diff_bytes.try_into().unwrap());
+            bytes_remaining = remainder;
+
+            dataz.push(((key, val), ts, diff as isize));
+        }
+        Ok(MemSnapshot { dataz })
+    }
+
+    fn allow_compaction(&mut self, _ts: u64) -> Result<(), Error> {
+        // No-op for now.
+        Ok(())
+    }
+}
+
+/// Implementation of [Persister]
+pub struct FilePersister {
+    registered: HashSet<Id>,
+    dataz: HashMap<Id, FileStream>,
+}
+
+impl FilePersister {
+    /// Constructs a new, empty SimplePersister.
+    pub fn new() -> Self {
+        FilePersister {
+            registered: HashSet::new(),
+            dataz: HashMap::new(),
+        }
+    }
+
+    /// Consumes this FilePersister, returning the underlying data.
+    /// TODO: I don't really get what these functions are meant to
+    /// illustrate I think its the action of "start a new dataflow to mock
+    /// out a restart?"
+    /// TODO: It seems like a want a "reload" method or something like that, in
+    /// a way that wouldn't make sense for a in-mem implementation
+    #[cfg(test)]
+    pub fn into_inner(self) -> HashMap<Id, FileStream> {
+        self.dataz
+    }
+
+    /// Constructs a FilePersister from a previous FilePersister's data but with
+    /// the create_or_load registrations reset.
+    #[cfg(test)]
+    pub fn from_inner(inner: HashMap<Id, FileStream>) -> Self {
+        FilePersister {
+            registered: HashSet::new(),
+            dataz: inner,
+        }
+    }
+}
+
+impl Persister for FilePersister {
+    type Write = FileStream;
+    type Meta = FileStream;
+
+    fn create_or_load(&mut self, id: Id) -> Result<Token<Self::Write, Self::Meta>, Error> {
+        if self.registered.contains(&id) {
+            return Err(format!("internal error: {:?} already registered", id).into());
+        }
+        self.registered.insert(id);
+        let p = self
+            .dataz
+            .entry(id)
+            .or_insert_with(|| FileStream::new(format!("test-{}", id.0)));
+        let t = Token {
+            write: p.clone(),
+            meta: p.clone(),
+        };
+        Ok(t)
+    }
+
+    // Note that we don't remove from registered here and maybe we ought to?
```

Comment:
I'm still dunno what I feel about the registrations at all. The intent was that a given process should only be able to hand out a writer for some id once. Still TODO is for the blob/buffer instances to implement their own process exclusivity locking. Combined, we'd be ensuring that at any time only one thing is ever writing to a stream at a time.

This clearly deserves a comment, but destroy intentionally doesn't remove from registered because it's already handed out that id.

Reply:
delete the note or replace it with some variant of my comment here plz